028-service-binding.md

```diff
@@ -0,0 +1,374 @@
+    namespace: my-namespace # optional
+status:
+  binding:
+    name: barista-kafka
+```
+
+`Secret` created by Strimzi:
+
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: barista-kafka
+  namespace: my-namespace
+type: Opaque
+  data:
+    type: kafka
+    provider: strimzi
+    bootstrapServer: # comma separated list of host:port
+  # Provided if TLS enabled:
+    ca.crt: #  Strimzi cluster CA certificate
+    ca.p12: #  PKCS #12 archive file for Strimzi cluster CA certificate
+    ca.password: # Password for protecting the Strimzi cluster CA certificate PKCS #12 archive file
+  # Provided if selected user is SCRAM auth:
+    username: # SCRAM username
+    password: # SCRAM password
+    sasl.jaas.config: # sasl jaas config string for use by Java applications
+  # Provided if selected user is mTLS:
+    user.p12: # client certificate for the consuming client PKCS #12 archive file for storing certificates and keys
+    user.password: # password for protecting the client certificate PKCS #12 archive file
+    user.crt: # certificate for the consuming client signed by the clients' CA
+    user.key: # private key for the consuming client
+```
+
+Naming of secret keys:
+
+The names for the keys given above match what is currently provided in other secrets created by Strimzi. The service binding specification does included some suggested fields for certificates, but they do not satisfy the requirements of an application connecting to Strimzi. These fields could be proposed back to the Service Binding specification.
+
+An alternative approach for naming is to name the fields to match the use, for example `ca` -> `truststore` (`truststore.crt`, `truststore.p12`, `truststore.password`), `user` -> `keystore` (`keystore.p12`, `keystore.password`, `keystore.crt`, `keystore.key`)
+ 
+
+### Implementation
+
+This proposal will result in a new `KafkaConnectionAssemblyOperator` class being added to the `cluster-operator` directory in the `strimzi-kafka-operator` repo. The class will be 
+deployed as a separate Verticle and have its own reconcile loop.
+
+The actions taken during the reconcile loop will be:
+1. On start-up the operator will watch all resources of Kind `KafkaConnection`.
+2. For any `KafkaConnection` resource present it will perform a GET request on the Kafka cluster that is 
+referenced in the `KafkaConnection` and pick out the configured listeners.
+3. If there are multiple listeners listed in the `Kafka` CR with the type `internal`, the operator 
+will filter the list by comparing the `tls` and `authentication` properties in the `Kafka` and `KafkaUser` CRs.
+4. If there are still multiple listeners it will pick one at random.
+5. The operator will add the name of the listener chosen will be added to the `KafkaConnection` status so that in future reconciles 
+the operator can take this into account in step 4 and avoid cycling through listeners.
+6. If there are no listeners marked as type `internal`, the reconcile loop will fail.
```

Comment:
So we can't bind to an external listener? Is there a good reason for that?

Reply:
I am curious about this as well ... why the bootstrap server in the Secret can't be one for an external listener?

Reply:
The reason I restricted it to an internal listener is because from a service binding point of view the application is expected to bind to the secret, so the application will be running in the cluster. Can operators watch resources in other Kubernetes clusters? If they can that would be a reason to open it up to external listeners as well

Reply:
I think we should have the following logic:
* Find all listeners with matching AuthN
* If more of them are found and one or more are internal, we should prefer the internal
* If only external is found with matching AuthN, we should take it